### PR TITLE
New Aggregate Healthcheck

### PR DESCRIPTION
**Pull request by kovacshuni**

### CODE REVIEW

File: aggregate-healthcheck-sidekick@.service [hunk 1]
```diff
@@ -7,7 +7,14 @@ After=aggregate-healthcheck@%i.service
 ExecStartPre=/bin/sh -c "\
   etcdctl set /vulcand/backends/aggregate-healthcheck/backend '{\"Type\": \"http\"}'; \
   etcdctl set /vulcand/frontends/aggregate-healthcheck/frontend '{\"Type\": \"http\", \"BackendId\": \"aggregate-healthcheck\", \"Route\": \"Host(`aggregate-healthcheck`)\"}'; \
-  etcdctl set /vulcand/frontends/aggregate-healthcheck-public/frontend '{\"Type\":\"http\", \"BackendId\":\"aggregate-healthcheck\", \"Route\":\"Path(`/__health`)\"}';"
+  etcdctl set /vulcand/frontends/aggregate-healthcheck-public/frontend '{\"Type\":\"http\", \"BackendId\":\"aggregate-healthcheck\", \"Route\":\"Path(`/__health`)\"}'; \
+  etcdctl set /vulcand/frontends/aggregate-healthcheck-public-gtg/frontend '{\"Type\":\"http\", \"BackendId\":\"aggregate-healthcheck\", \"Route\":\"Path(`/__gtg`)\"}'; \
+  etcdctl set /ft/healthcheck-categories/default/period_seconds 60; \
```
Comment (Zapadlo) [inline aggregate-healthcheck-sidekick@.service]: Unsure about the value of these, are they meant to be defaults that can be overwritten in etcd? They will reset every time service restarts though

File: annotations-api-sidekick@.service [hunk 2]
```diff
@@ -14,7 +14,8 @@ ExecStart=/bin/sh -c "\
   \
   while true; do \
     export ANNOTATIONS_API_PORT=$(echo $(/usr/bin/docker port annotations-api-%i 8080) | cut -d':' -f2); \
-    etcdctl set /services/api-policy-component/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/api-policy-component/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/api-policy-component/%i/healthcheck/categories default,read --ttl 60; \
```
Comment (Zapadlo) [inline annotations-api-sidekick@.service]: Does the category really need to be updated every 45s?

File: organisation-read-api-sidekick@.service [hunk 3]
```diff
@@ -15,7 +15,8 @@ ExecStart=/bin/sh -c "\
   \
   while true; do \
     export ORGANISATION_READ_API_PORT=$(echo $(/usr/bin/docker port organisation-read-api-%i 8080) | cut -d':' -f2); \
-    etcdctl set /services/organisation-read-api/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/organisation-read-api/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/organisation-read-api/%i/healthcheck/categories default,read --ttl 60; \
```
Comment (Zapadlo) [inline organisation-read-api-sidekick@.service]: Again, inside the loop

File: people-read-api-sidekick@.service [hunk 4]
```diff
@@ -15,7 +15,8 @@ ExecStart=/bin/sh -c "\
   \
   while true; do \
     export PEOPLE_READ_API_PORT=$(echo $(/usr/bin/docker port people-read-api-%i 8080) | cut -d':' -f2); \
-    etcdctl set /services/people-read-api/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/people-read-api/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/people-read-api/%i/healthcheck/categories default,read --ttl 60; \
```
Comment (Zapadlo) [inline people-read-api-sidekick@.service]: Here

File: semantic-reader-sidekick@.service [hunk 5]
```diff
@@ -17,7 +17,8 @@ ExecStart=/bin/sh -c "\
     etcdctl set /services/semantic-reader/%i/host '%H' --ttl 60; \
     etcdctl set /services/semantic-reader/%i/app_port $(echo $(/usr/bin/docker port semantic-reader-%i 8080) | cut -d':' -f2) --ttl 60; \
     etcdctl set /services/semantic-reader/%i/admin_port $(echo $(/usr/bin/docker port semantic-reader-%i 8081) | cut -d':' -f2) --ttl 60; \
-    etcdctl set /services/semantic-reader/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/semantic-reader/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/semantic-reader/%i/healthcheck/categories default --ttl 60; \
```
Comment (Zapadlo) [inline semantic-reader-sidekick@.service]: Here

File: wp-notifier-sidekick@.service [hunk 6]
```diff
@@ -16,7 +16,8 @@ ExecStart=/bin/sh -c "\
     etcdctl set /services/wp-notifier/%i/host '%H' --ttl 60; \
     etcdctl set /services/wp-notifier/%i/app_port $(echo $(/usr/bin/docker port wp-notifier-%i 8080) | cut -d':' -f2) --ttl 60; \
     etcdctl set /services/wp-notifier/%i/admin_port $(echo $(/usr/bin/docker port wp-notifier-%i 8081) | cut -d':' -f2) --ttl 60; \
-    etcdctl set /services/wp-notifier/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/wp-notifier/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/wp-notifier/%i/healthcheck/categories default --ttl 60; \
```
Comment (Zapadlo) [inline wp-notifier-sidekick@.service]: Here

File: wp-transformer-sidekick@.service [hunk 7]
```diff
@@ -16,7 +16,8 @@ ExecStart=/bin/sh -c "\
     etcdctl set /services/wp-transformer/%i/host '%H' --ttl 60; \
     etcdctl set /services/wp-transformer/%i/app_port $(echo $(/usr/bin/docker port wp-transformer-%i 8080) | cut -d':' -f2) --ttl 60; \
     etcdctl set /services/wp-transformer/%i/admin_port $(echo $(/usr/bin/docker port wp-transformer-%i 8081) | cut -d':' -f2) --ttl 60; \
-    etcdctl set /services/wp-transformer/%i/healthcheck /__health --ttl 60; \
+    etcdctl set /services/wp-transformer/%i/healthcheck/path /__health --ttl 60; \
+    etcdctl set /services/wp-transformer/%i/healthcheck/categories default --ttl 60; \
```
Comment (Zapadlo) [inline wp-transformer-sidekick@.service]: And here

File: api-policy-component-sidekick@.service [hunk 8]
```diff
@@ -7,7 +7,8 @@ After=api-policy-component@%i.service
 ExecStart=/bin/sh -c "\
   etcdctl set /ft/services/api-policy-component/healthcheck true; \
   etcdctl mkdir /ft/services/api-policy-component/servers; \
-  etcdctl set /ft/healthcheck/api-policy-component-%i /__health; \
+  etcdctl set /ft/healthcheck/api-policy-component-%i/path /__health; \
```
Comment (catalin-ilea) [inline api-policy-component-sidekick@.service]: Isn't this redundant? Is the health path ever going to be different?

File: binary-ingester-sidekick@.service [hunk 9]
```diff
@@ -14,11 +13,14 @@ ExecStart=/bin/sh -c "\
   \
   etcdctl set /vulcand/frontends/binary-ingester-health-%i/frontend \"{\\\"Type\\\":\\\"http\\\", \\\"BackendId\\\":\\\"binary-ingester-%i\\\", \\\"Route\\\":\\\"Path(\`/health/binary-ingester-%i/__health\`)\\\"}\"; \
   etcdctl set /vulcand/frontends/binary-ingester-health-%i/middlewares/rewrite \"{\\\"Id\\\":\\\"rewrite\\\", \\\"Type\\\":\\\"rewrite\\\", \\\"Priority\\\":1, \\\"Middleware\\\": {\\\"Regexp\\\":\\\"/health/binary-ingester-%i(.*)\\\", \\\"Replacement\\\":\\\"\$1\\\"}}\"; \
-  etcdctl set /ft/healthcheck/binary-ingester-%i /__health; \
-  \
+  etcdctl set /ft/healthcheck/binary-ingester-%i/path /__health; \
+  etcdctl set /ft/healthcheck/binary-ingester-%i/categories default; \
```
Comment (catalin-ilea) [inline binary-ingester-sidekick@.service]: Should we assume default if the key is missing? This way you'd avoid a big-bang release where all services will get restarted.

Comment (Zapadlo) [inline binary-ingester-sidekick@.service]: +1

Comment (Zapadlo) [inline binary-ingester-sidekick@.service]: +1

**Comment by kovacshuni** [comment]
Okay, so we talked about the etcd keys with Catalin.
I'm removing categories where not necessary, the absence will mean default.
We tried to do something to not need the modification for the path either but that can't be done, as the key /ft/healthcheck/service-name-1 is now a directory and before it was a leaf node. Categories wouldn't have a place to be in if it stayed a leaf node. So remained a big-bang-ish release for now. Hope it goes well.

**Comment by kovacshuni** [comment]
I updated.
Default category is not even defined, and not assigned, it will be part of the agg-health program to associate with all services and its time will be 60 seconds from code and non-resilient.
Extracted the lines from while loops to the outside.
The .../path __health i had to leave, that marks a service that it is defined and has to be checked upon.

**Comment by tosan88** [comment]
:+1:

**Comment by Zapadlo** [comment]
+1

**Comment by catalin-ilea** [comment]
Can you please double check you modified all services which are deployed and expose HCs? You're missing topics-rw-neo4j...might be others as well

**Comment by kovacshuni** [comment]
I merged latest master in here. Fixed up topics-rw-neo4j, this was a new one. Double checked all services have their etcd health key correct. Testin in xp.

**Comment by catalin-ilea** [comment]
Brill

**Comment by kovacshuni** [comment]
```
Agg-hc deployment plan

#clean-up old etcd values, right after merge

1. etcdctl rm --recursive /ft/healthcheck
2. fleetctl list-units | grep sidekick | awk '{print $1}' | xargs fleetctl destroy
```

**Comment by kovacshuni** [comment]
* Tested in XP - worked like a charm
* Tested in pre-prod. - worked like a charm

**Pull request merged**